libtenzir/builtins/operators/tcp-listen.cpp

```diff
@@ -0,0 +1,512 @@
+  }
+
+  auto diagnostics() noexcept -> diagnostic_handler& override {
+    return diagnostics_;
+  }
+
+  auto no_location_overrides() const noexcept -> bool override {
+    return no_location_overrides_;
+  }
+
+  auto has_terminal() const noexcept -> bool override {
+    return has_terminal_;
+  }
+
+  auto set_waiting(bool value) noexcept -> void override {
+    (void)value;
+    TENZIR_UNIMPLEMENTED();
+  }
+
+private:
+  shared_diagnostic_handler diagnostics_;
+  bool has_terminal_;
+  bool no_location_overrides_;
+};
+
+using bridge_actor = caf::typed_actor<
+  // Forwards slices from the connection actors to the operator
+  auto(table_slice slice)->caf::result<void>,
+  auto(atom::get)->caf::result<table_slice>>;
+
+struct connection_state {
+  ~connection_state() noexcept {
+    if (tls_socket) {
+      tls_socket->shutdown();
+      tls_socket->lowest_layer().close();
+    } else if (socket) {
+      socket->close();
+    }
+  }
+
+  caf::event_based_actor* self;
+  std::optional<boost::asio::ip::tcp::socket> socket = {};
+  std::optional<boost::asio::ssl::context> ssl_ctx = {};
+  std::optional<boost::asio::ssl::stream<boost::asio::ip::tcp::socket&>>
+    tls_socket = {};
+  detail::weak_handle<bridge_actor> bridge;
+  tcp_listen_args args;
+  std::unique_ptr<operator_control_plane> ctrl;
+
+  generator<table_slice> gen = {};
+  generator<table_slice>::iterator it = {};
+};
+
+auto make_connection(caf::stateful_actor<connection_state>* self,
+                     boost::asio::ip::tcp::socket socket, bridge_actor bridge,
+                     tcp_listen_args args,
+                     shared_diagnostic_handler diagnostics) -> caf::behavior {
+  self->state.self = self;
+  self->state.socket = std::move(socket);
+  self->state.bridge = std::move(bridge);
+  self->state.args = std::move(args);
+  self->state.ctrl = std::make_unique<tcp_listen_control_plane>(
+    std::move(diagnostics), args.has_terminal, args.no_location_overrides);
+  if (self->state.args.tls) {
+    self->state.ssl_ctx.emplace(boost::asio::ssl::context::tls_client);
+    self->state.ssl_ctx->set_default_verify_paths();
+    self->state.ssl_ctx->set_verify_mode(
+      boost::asio::ssl::verify_peer
+      | boost::asio::ssl::verify_fail_if_no_peer_cert);
+    self->state.tls_socket.emplace(*self->state.socket, *self->state.ssl_ctx);
+    auto tls_handle = self->state.tls_socket->native_handle();
+    if (SSL_set1_host(tls_handle, self->state.args.hostname.c_str()) != 1) {
+      diagnostic::error("failed to enable host name verification")
+        .emit(self->state.ctrl->diagnostics());
+      return {};
+    }
+    if (not SSL_set_tlsext_host_name(tls_handle,
+                                     self->state.args.hostname.c_str())) {
+      diagnostic::error("failed to set SNI")
+        .emit(self->state.ctrl->diagnostics());
+      return {};
+    }
+    auto ec = boost::system::error_code{};
+    self->state.ssl_ctx.emplace(boost::asio::ssl::context::tls_server);
```

Comment:
Aren't we unconditionally overwriting the previous context from line 136 here?

Reply:
Yes, but I think we may actually have the same bug in `tcp.cpp` then because that's what this is copied from.

Reply:
Let's take the opportunity to fix it in both places then :D